File: 1816-lowest-common-ancestor-of-a-binary-tree-iv/1816-lowest-common-ancestor-of-a-binary-tree-iv.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {

    private Set<TreeNode> targetNodes;

    public TreeNode lowestCommonAncestor(TreeNode root, TreeNode[] nodes) {

        if (root == null) {
            return null;
        }

        targetNodes = new HashSet<>();
        for (TreeNode node : nodes) {
            targetNodes.add(node);
        }

        return findLCA(root);

    }

    private TreeNode findLCA(TreeNode root) {
        if (root == null) {
            return null;
        }
        if (this.targetNodes.contains(root)) {
            return root;
        }

        TreeNode left = findLCA(root.left);
        TreeNode right = findLCA(root.right);

        if (left != null && right != null) {
            return root;
        }

        return (left != null) ? left : right;
    }
}

File: 1816-lowest-common-ancestor-of-a-binary-tree-iv/1816-lowest-common-ancestor-of-a-binary-tree-iv.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def lowestCommonAncestor(self, root: 'TreeNode', nodes: 'List[TreeNode]') -> 'TreeNode':

        memo = set()
        for node in nodes:
            memo.add(node)
        
        def findLCA(node: 'TreeNode') -> 'TreeNode':
            if not node or node in memo:
                return node
            
            left = findLCA(node.left)
            right = findLCA(node.right)

            if left and right:
                return node
            
            return left if left else right

        return findLCA(root)
        